### PR TITLE
first working version

### PR DESCRIPTION
**Pull request by romintomasetti**
This PR adds:
- the package
- some CI
- a devcontainer

### CODE REVIEW

**Comment by maartenarnst** [comment]
OK. I've taken a look at the PR. It all looks good. From the bigger picture point of view, there's the question of whether we want to extract from the `tools` in `trilinos-container` separate pieces like we do here for the `apt` helpers. Or whether we'd want to extract the entire `tools`. It's a bit the question of how all the helpers would be most readable to the "outside" world: several small repos vs one larger repo. Thinking about it the other way around, we could also think whether it may make sense to suggest these helpers as features in an issue on the `apt` GitHub. Keeping it as is here would be OK for me, but probably best discussed a bit live.

File: system_helpers/update_alternatives/script.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+@typeguard.typechecked
+def parse_args() -> argparse.Namespace:
+    """
+    Parse CLI arguments.
+    """
+    parser = argparse.ArgumentParser()
+
+    subparsers = parser.add_subparsers(required = True)
+
+    parser_ua = subparsers.add_parser('update-alternatives')
```
Comment (romintomasetti) [inline system_helpers/update_alternatives/script.py]: We should not use a sub-parser, because it's now a standalone script for `update-alternatives`.
Usage should be:
```bash
update-alternative-helpers <args>
```
instead of current:
```bash
update-alternatives-helpers update-alternatives <args>
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            image: python:${{ matrix.version }}
+        strategy:
+            matrix:
+                version: ['3.10', '3.12']
+        steps:
+            - name: Install as package.
+              run : |
+                  pip install git+https://github.com/uliegecsm/system-helpers.git@${{ github.sha }}
+
+            - name: Test 'apt' helpers as CLI directly.
```
Comment (maartenarnst) [inline .github/workflows/test.yml]:
```suggestion
            - name: Test 'apt' helpers CLI directly.
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+              run : |
+                  apt-helpers install-packages --update --clean --upgrade --packages jq
+
+            - name: Test 'update-alternatives' helpers as CLI directly.
```
Comment (maartenarnst) [inline .github/workflows/test.yml]:
```suggestion
            - name: Test 'update-alternatives' helpers CLI directly.
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            matrix:
+                version: ['3.10', '3.12']
+        steps:
+            - name: Install as package.
+              run : |
+                  pip install git+https://github.com/uliegecsm/system-helpers.git@${{ github.sha }}
+
+            - name: Test 'apt' helpers as CLI directly.
+              run : |
+                  apt-helpers install-packages --update --clean --upgrade --packages jq
```
Comment (maartenarnst) [inline .github/workflows/test.yml:42]: Should this be followed by a test of some call of `jq`? Could be just asking to print out its help. The idea would be to make sure that it's installed.

File: system_helpers/apt/script.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        '--packages',
+        help = "List of packages to install.",
+        nargs = '*',
+        required = False,
+        dest = 'packages',
+    )
+
+    parser_ip.add_argument(
+        '--requirement',
+        help = 'Requirement file à la pip.',
```
Comment (maartenarnst) [inline system_helpers/apt/script.py:34]:
```suggestion
        '--requirements',
        help = 'Requirements file à la pip.',
```

File: system_helpers/update_alternatives/script.py (hunk 1 above)
Comment (maartenarnst) [reply system_helpers/update_alternatives/script.py]: Indeed. It's probably the case for the "install packages" too.

File: tests/apt/test_install.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+@pytest.fixture
+@typeguard.typechecked
+def requirements() -> typing.Generator[typing.Tuple[typing.List[pathlib.Path], typing.List[str]], None, None]:
+    """
+    Get requirements files *à la* `pip`.
+    """
+    with tempfile.NamedTemporaryFile(mode = 'w+') as req_1, \
+         tempfile.NamedTemporaryFile(mode = 'w+') as req_2:
+        req_1.write("# Let's start this first requirement file with a comment. Then, add some packages.\n")
```
Comment (maartenarnst) [inline tests/apt/test_install.py]:
```suggestion
        req_1.write("# Let's start this first requirements file with a comment. Then, add some packages.\n")
```

File: tests/apt/test_install.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        req_1.write("git\n")
+        req_1.write("and\n")
+        req_1.write("\n")
+        req_1.write("# Some other useless comments here.\n")
+        req_1.write("whatnot\n")
+        req_1.flush()
+
+        req_2.write("# Let's start this other requirement file with a comment. Then, add some packages.\n")
```
Comment (maartenarnst) [inline tests/apt/test_install.py]:
```suggestion
        req_2.write("# Let's start this other requirements file with a comment. Then, add some packages.\n")
```

File: tests/apt/test_install.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+        with unittest.mock.patch(target = 'subprocess.check_call', side_effect = [None, None, None]) as mocker:
+            install.install_packages(packages = packages, update = False, upgrade = True, clean = False)
+
+            mocker.assert_has_calls(calls = [
+                unittest.mock.call(['apt', '--yes', 'upgrade']),
+                unittest.mock.call(['apt', '--yes', '--no-install-recommends', 'install'] + packages),
+            ])
+
+    def test_list_of_requirement_files(self, requirements):
```
Comment (maartenarnst) [inline tests/apt/test_install.py]:
```suggestion
    def test_list_of_requirements_files(self, requirements):
```
You do a search replace? :)

File: tests/apt/test_install.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        with unittest.mock.patch(target = 'subprocess.check_call', side_effect = [None]) as mocker:
+            install.install_packages(requirements = requirements[0], update = False, upgrade = False, clean = False)
+
+            mocker.assert_has_calls(calls = [
+                unittest.mock.call(['apt', '--yes', '--no-install-recommends', 'install'] + requirements[1]),
+            ])
+
+    @unittest.mock.patch(target = 'subprocess.check_call', side_effect = [None])
+    @pytest.mark.script_launch_mode('inprocess')
+    def test_install_packages_from_cli(self, mocker, script_runner : pytest_console_scripts.ScriptRunner, requirements):
```
Comment (maartenarnst) [inline tests/apt/test_install.py:78]: OK. That's nice!

Comment (maartenarnst) [reply tests/apt/test_install.py:78]: May be worth double checking if the `side_effect = [None]` is needed. May also be worth it for the next test where you use `= 6 * [None]`.

File: system_helpers/apt/script.py (hunk 5 above)
Comment (romintomasetti) [reply system_helpers/apt/script.py:34]: It mimics what `pip` expects:
```bash
  -r, --requirement <file>    Install from the given requirements file. This option can be used multiple times.
```

File: system_helpers/update_alternatives/script.py (hunk 1 above)
Comment (romintomasetti) [reply system_helpers/update_alternatives/script.py]: Well, for `apt-helpers` I would still keep the `install` subparser, because `apt` has many other actions possible.

File: .github/workflows/test.yml (hunk 2 above)
Comment (romintomasetti) [reply .github/workflows/test.yml]: :heavy_check_mark:

File: .github/workflows/test.yml (hunk 3 above)
Comment (romintomasetti) [reply .github/workflows/test.yml]: :heavy_check_mark:

File: .github/workflows/test.yml (hunk 4 above)
Comment (romintomasetti) [reply .github/workflows/test.yml:42]: :heavy_check_mark:

File: tests/apt/test_install.py (hunk 6 above)
Comment (romintomasetti) [reply tests/apt/test_install.py]: :heavy_check_mark:

File: tests/apt/test_install.py (hunk 7 above)
Comment (romintomasetti) [reply tests/apt/test_install.py]: :heavy_check_mark:

File: tests/apt/test_install.py (hunk 8 above)
Comment (romintomasetti) [reply tests/apt/test_install.py]: :heavy_check_mark:

File: tests/apt/test_install.py (hunk 9 above)
Comment (romintomasetti) [reply tests/apt/test_install.py:78]: Discussed live. This serves as a "how many times it should be mocked" test. If the mocker is called more times than the length of the `side_effect` list, you'd get a `StopIteration` exception.

File: system_helpers/apt/install.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    cmd = ['apt']
+
+    if yes:
+        cmd.append('--yes')
+
+    if no_install_recommends:
+        cmd.append('--no-install-recommends')
+
+    cmd.append('install')
+    return cmd
```
Comment (maartenarnst) [inline system_helpers/apt/install.py:40]:
```suggestion
    return cmd.append('install')
```

File: system_helpers/apt/install.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    to_be_installed = []
+
+    if packages:
+        to_be_installed += packages
+
+    if requirements:
+        for file in requirements:
+            to_be_installed += get_list_of_packages_from_requirements_file(file = file)
+
+    logging.info(f"Installing 'apt' packages {to_be_installed} (update={update}, upgrade={upgrade}, clean={clean})")
```
Comment (maartenarnst) [inline system_helpers/apt/install.py:70]: Should there be a check that `to_be_installed` is not empty with a "raise" if it is empty?

File: system_helpers/apt/install.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    if update:
+        subprocess.check_call(['apt', 'update'])
+
+    if upgrade:
+        subprocess.check_call(['apt', '--yes', 'upgrade'])
+
+    args = install_command(yes = True, no_install_recommends = True)
+    args += to_be_installed
+
+    subprocess.check_call(args)
```
Comment (maartenarnst) [inline system_helpers/apt/install.py:81]: Just wondering if there should be a way to set the `DEBIAN_FRONTEND=noninteractive` environment variable. It could also be left to the user to do that before calling the script.

File: tests/apt/test_install.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        return pathlib.Path(__file__).parent.parent.parent / 'system_helpers' / 'apt' / 'script.py'
+
+    def test_list_of_package_names(self):
+        """
+        Test for a provided list of package names.
+        """
+        packages = ['git', 'and', 'whatnot']
+
+        with unittest.mock.patch(target = 'subprocess.check_call', side_effect = [None, None, None]) as mocker:
```
Comment (maartenarnst) [inline tests/apt/test_install.py]: Would it be enough to have just `[None, None]`? :phone:

File: system_helpers/update_alternatives/script.py (hunk 1 above)
Comment (maartenarnst) [reply system_helpers/update_alternatives/script.py]: OK. I agree!

**Comment by maartenarnst** [comment]
OK. I've just read through the entire PR once more. It's very nice! Just a few details left.

File: system_helpers/apt/install.py (hunk 10 above)
Comment (romintomasetti) [reply system_helpers/apt/install.py:40]: `append` does not return anything.

File: system_helpers/apt/install.py (hunk 11 above)
Comment (romintomasetti) [reply system_helpers/apt/install.py:70]: It is legal to just do `apt install`.

Comment (romintomasetti) [reply system_helpers/apt/install.py:70]: But in this case, since we explicitly want packages to be installed, we can indeed enforce there is at least one.

Comment (romintomasetti) [reply system_helpers/apt/install.py:70]: Done.

File: system_helpers/apt/install.py (hunk 12 above)
Comment (romintomasetti) [reply system_helpers/apt/install.py:81]: Yep, let's not do it for now.

File: tests/apt/test_install.py (hunk 13 above)
Comment (romintomasetti) [reply tests/apt/test_install.py]: Indeed ! Done :heavy_check_mark:

**Review by maartenarnst — APPROVED** [review]
OK! 🟢

File: .devcontainer/devcontainer.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
+{
+    "dockerFile": "dockerfile",
+    "context": "..",
```
Comment (maartenarnst) [inline .devcontainer/devcontainer.json:3]: Just wondering if we need to set the "context"? I think we don't do it in our other repos.

**Review by maartenarnst — APPROVED** [review]
OK! Still good :) Just the thought about the "context" in the `devcontainer.json`.

File: .devcontainer/devcontainer.json (hunk 14 above)
Comment (romintomasetti) [reply .devcontainer/devcontainer.json:3]: The default context is the directory of the `dockerfile` (in this case, it is thus `.devcontainer/`).

But I need to mount the `requirements` directory.

So yes, it's needed in this case.

Comment (maartenarnst) [reply .devcontainer/devcontainer.json:3]: Ok! Got it!

**Comment by maartenarnst** [comment]
I’ll let you merge.
